schemachanger: add support discarding database/table zc

annrpom opened the pull request:
### scbuild: version gate SetZoneConfig with isV243Active
New support was added in 24.3 in `SetZoneConfig` for subzone
configs. Since this work was a significant change, we should
gate `SetZoneConfig` to minimize the risk of compatibility
issues with earlier versions.

Epic: none

Release note: None

---

### schemachanger: re-enable dsc zone config
Since 24.3 has been cut, we can turn zone configs
on in the declarative schema changer by default.

Epic: none

Release note: None

---

### scexec: extend UpdateZoneConfig to allow for deletes
This patch extends UpdateZoneConfig to allow for deletes in
the system.zones table.

Epic: None

Release note: None

---

### scbuildstmt: refactor zoneConfigObjBuilder method
This patch refactors the `addZoneConfigToBuildCtx` method
of our `zoneConfigObjBuilder` to be less specific; we pull
the `b.add` out so that we can decide in `SetZoneConfig`
whether to add or drop this element.

Epic: none

Release note: None

---

### schemachanger: add support discarding database/table zc
This patch supports discarding a database/table zone config
in the declarative schema changer.

Informs: https://github.com/cockroachdb/cockroach/issues/133157

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/132967)
<!-- Reviewable:end -->

annrpom committed:
schemachanger: re-enable dsc zone config

Since 24.3 has been cut, we can turn zone configs
on in the declarative schema changer by default.

Epic: none

Release note: None

annrpom committed:
scexec: extend UpdateZoneConfig to allow for deletes

This patch extends UpdateZoneConfig to allow for deletes in
the system.zones table.

Epic: None

Release note: None

annrpom committed:
scbuildstmt: refactor zoneConfigObjBuilder method

This patch refactors the `addZoneConfigToBuildCtx` method
of our `zoneConfigObjBuilder` to be less specific; we pull
the `b.add` out so that we can decide in `SetZoneConfig`
whether to add or drop this element.

Epic: none

Release note: None

annrpom committed:
schemachanger: add support discarding database/table zc

This patch supports discarding a database/table zone config
in the declarative schema changer.

Informs: #133157

Release note: None

rafiss reviewed (changes requested):
does the existing `zone_config` logic test already cover DISCARD well enough?

also, do we need a new end-to-end or scplan/scbuild test in the `schemachanger` package? i didn't see any existing ones there for adding zone configs. should we add tests for both adding and discarding zone configs there?

also, for `pkg/sql/schemachanger/scbuild/testdata/unimplemented_zone_cfg` does unskipping that require removing all zone config fallbacks, or could we do it now?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-:-O9z3uGr4kPoye0chxX_:bqqqnf0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/event_log.go` line 469 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z2F5Z-KtDugNY5CnL:-O9z2F5_6JKanzye7VoP:bgrxri) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/event_log.go#L469)):*
> ```Go
> 			var zcDetails eventpb.CommonZoneConfigDetails
> 			var oldConfig string
> 			if pb.maybePayload != nil {
> ```

can we avoid the unchecked interface conversion here and do:
```
if payload, ok := pb.maybePayload.(*eventpb.SetZoneConfig); ok {
  zcDetails = ...
  oldConfig = ...
}
```
___
*[`pkg/sql/schemachanger/scbuild/event_log.go` line 483 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z2Upl9PCpLrAFS-UA:-O9z2Upl9PCpLrAFS-UB:b-8nes5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/event_log.go#L483)):*
> ```Go
> 		} else {
> 			var zcDetails eventpb.CommonZoneConfigDetails
> 			if pb.maybePayload != nil {
> ```

can we avoid the unchecked interface conversion here and do:
```
if payload, ok := pb.maybePayload.(*eventpb.RemoveZoneConfig); ok {
  zcDetails = ...
  oldConfig = ...
}
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z-msW5IdGSv90ldro:-O9z-msW5IdGSv90ldrp:bxh4ndo) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbeabfa0970f11b37f448e7c7c3049a7035d1175/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L66)):*
> ```Go
> 	reflect.TypeOf((*tree.CreateSequence)(nil)):      {fn: CreateSequence, statementTags: []string{tree.CreateSequenceTag}, on: true, checks: isV241Active},
> 	reflect.TypeOf((*tree.CreateDatabase)(nil)):      {fn: CreateDatabase, statementTags: []string{tree.CreateDatabaseTag}, on: true, checks: isV241Active},
> 	reflect.TypeOf((*tree.SetZoneConfig)(nil)):       {fn: SetZoneConfig, statementTags: []string{tree.ConfigureZoneTag}, on: true, checks: isV242Active},
> ```

should the check be `isV243Active`?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 171 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z2ws-DcLMd5nz1NPz:-O9z2ws-DcLMd5nz1NQ-:b-th4c8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L171)):*
> ```Go
> 		needToError := false
> 		// Determine if this zone config that we're trying to discard is
> 		// supposed to be there. We are either a database or a table.
> ```

nit: does "we" here refer to `zco`? i think it would be more clear if the comment was phrased that way

also, why do we know the zone config is only for a DB or table at this point?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 241 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z3So86q3F8d3PzUmP:-O9z3So86q3F8d3PzUmQ:bv2fup8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L241)):*
> ```Go
> 	} else {
> 		return false, errors.AssertionFailedf(
> 			"unknown table locality")
> ```

could there be more information in this error message that says what the elements were, so we can see which field was unknown?


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> does the existing `zone_config` logic test already cover DISCARD well enough?

- it tests that discarding a zone config that does not exist (in system.zones) does nothing. one thing to note is that in the declarative schema changer, it actually does nothing (we no-op). the legacy schema changer will log a discard event
- it tests that we block discard for multi-region entities
- it tests that discarding a table zone config with subzone configs on it will still leave the subzone configs in-tact

i feel good about this coverage, but let me know if you feel differently

> also, do we need a new end-to-end or scplan/scbuild test in the `schemachanger` package? i didn't see any existing ones there for adding zone configs. should we add tests for both adding and discarding zone configs there?

i'll add some end-to-end tests for discard (i think i dreamed of doing this last night instead of actually doing it); we have some for the add direction: https://github.com/cockroachdb/cockroach/blob/ed42141682cf9536cda4b417762255b5038ad7f9/pkg/sql/schemachanger/testdata/end_to_end/alter_table_configure_zone_multiple/alter_table_configure_zone_multiple.definition (for example)

re: `pkg/sql/schemachanger/scbuild/testdata/unimplemented_zone_cfg` -- let me check

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-:-O9z5pbIBGkOtJe5FkPS:bc63bic)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/event_log.go` line 469 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z2F5Z-KtDugNY5CnL:-O9z9FfNCqXxoK9fmOV-:bq77hyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/event_log.go#L469)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can we avoid the unchecked interface conversion here and do:
```
if payload, ok := pb.maybePayload.(*eventpb.SetZoneConfig); ok {
  zcDetails = ...
  oldConfig = ...
}
```
</blockquote></details>

oki done
___
*[`pkg/sql/schemachanger/scbuild/event_log.go` line 483 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z2Upl9PCpLrAFS-UA:-O9zFsjWECjvvP9BZr6b:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/event_log.go#L483)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can we avoid the unchecked interface conversion here and do:
```
if payload, ok := pb.maybePayload.(*eventpb.RemoveZoneConfig); ok {
  zcDetails = ...
  oldConfig = ...
}
```
</blockquote></details>

done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z-msW5IdGSv90ldro:-O9zFvcQB3t2Wv-3ZluU:btukek1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbeabfa0970f11b37f448e7c7c3049a7035d1175/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L66)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should the check be `isV243Active`?
</blockquote></details>

i figured it would be 24.2 since we started supporting db/table zone config in that version; let me know if this understanding is incorrect
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 171 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z2ws-DcLMd5nz1NPz:-O9zGMKsDB0wq2LVo3pk:b-7v87dz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L171)):*
it does; i changed the phrasing!

> also, why do we know the zone config is only for a DB or table at this point?

this function gets called for either a database or table zone configuration object; a table zone configuration object includes index/partition subzone configs since those are "apart" of the zone config
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 241 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z3So86q3F8d3PzUmP:-O9zHiZx6PAWo8I7qmSH:b-49ekkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b172e554a8e68959d912a3710e223d9176b387bc/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L241)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could there be more information in this error message that says what the elements were, so we can see which field was unknown?
</blockquote></details>

let me see how i can do this -- i think printing out all of the elements would be a lot; this error check was taken from the legacy schema changer where we could just throw in the table descriptor's locality config


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
let me correct/be more precise on:

> for multi-region entities

i meant for multi-region zcs created by multi-region abstractions

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-:-OA-6CYC5ENwPNDl2U2p:b-4kp7lk)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

annrpom commented:
> re: pkg/sql/schemachanger/scbuild/testdata/unimplemented_zone_cfg -- let me check

looks like unskipping requires removing fallbacks

rafiss reviewed (approved):
lgtm! just had a comment on the version gate

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-:-OAOSipc36ocWL_3Hkcb:bp13rf6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/132967#-O9z-msW5IdGSv90ldro:-OAOSOys4tUVitpwbC-x:b4y1p8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbeabfa0970f11b37f448e7c7c3049a7035d1175/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/process.go#L66)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

i figured it would be 24.2 since we started supporting db/table zone config in that version; let me know if this understanding is incorrect
</blockquote></details>

this should be 24.3, since there was development work in between 24.2 and 24.3, and not all of the related PRs were backported to 24.3


<!-- Sent from Reviewable.io -->

annrpom committed:
scbuild: version gate SetZoneConfig with isV243Active

New support was added in 24.3 in `SetZoneConfig` for subzone
configs. Since this work was a significant change, we should
gate `SetZoneConfig` to minimize the risk of compatibility
issues with earlier versions.

Epic: none

Release note: None

annrpom commented:
TFTR! ('-')7 

bors r=rafiss
